File: openmole/plugins/org.openmole.plugin.environment.glite/src/main/java/org/openmole/plugin/environment/glite/GliteJob.java
/*
 *  Copyright (C) 2010 reuillon
 * 
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.plugin.environment.glite;

import java.util.logging.Level;
import java.util.logging.Logger;
import org.openmole.commons.exception.InternalProcessingError;
import org.openmole.core.model.execution.ExecutionState;
import org.openmole.plugin.environment.jsaga.JSAGAJob;

/**
 *
 * @author reuillon
 */
public class GliteJob extends JSAGAJob {

    final static Logger LOGGER = Logger.getLogger(GliteJob.class.getName());
    
    final long proxyExpired;
    
    public GliteJob(String jobId, GliteJobService jobService, long proxyExpired) throws InternalProcessingError {
        super(jobId, jobService);
        this.proxyExpired = proxyExpired;
        LOGGER.log(Level.FINE, "Job will be killed if not executed before {0}", proxyExpired);
    }

    @Override
    public ExecutionState updateState() throws InternalProcessingError {
        if(proxyExpired < System.currentTimeMillis()) throw new InternalProcessingError("Proxy for this job has expired.");
        return super.updateState();
    }

    
    
    
}

File: openmole/plugins/org.openmole.plugin.environment.glite/src/main/scala/org/openmole/plugin/environment/glite/GliteJob.scala
/*
 * Copyright (C) 2012 reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.plugin.environment.glite

import org.openmole.core.batch.environment._
import org.openmole.core.batch.control.AccessToken
import org.openmole.core.batch.jobservice.{ BatchJob, BatchJobId }
import org.openmole.core.model.execution.ExecutionState._
import org.openmole.misc.workspace._

trait GliteJob extends BatchJob with BatchJobId {
  var lastShaked = System.currentTimeMillis

  override def updateState(implicit token: AccessToken) = {
    val state = super.updateState

    //if (!state.isFinal && proxyExpired < System.currentTimeMillis) throw new InternalProcessingError("Proxy for this job has expired.")

    if (state == SUBMITTED) {
      val maxNbReady = Workspace.preferenceAsInt(GliteEnvironment.JobShakingMaxReady)

      def nbReady = jobService.environment.jobRegistry.allExecutionJobs.count(_.state == READY)

      if (nbReady < maxNbReady) {

        val jobShakingAverageTime = Workspace.preferenceAsDurationInMs(GliteEnvironment.JobShakingHalfLife)
        val nbInterval = ((System.currentTimeMillis - lastShaked.toDouble) / jobShakingAverageTime)
        val probability = 1 - math.pow(0.5, nbInterval)

        lastShaked = System.currentTimeMillis

        if (Workspace.rng.nextDouble < probability) throw new ShouldBeKilledException("Killed in shaking process")
      }
    }
    state
  }
}
